### PR TITLE
Concept identification pointer generator

### PR DESCRIPTION
**Pull request by teodora99**

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
```
Comment (Ranapop) [inline .idea/.gitignore]: why is there another .gitignore -- remove this

File: .idea/amr-parsing.iml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Ranapop) [inline .idea/amr-parsing.iml]: ignore this file (probably all .idea)

File: .idea/inspectionProfiles/profiles_settings.xml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<component name="InspectionProjectProfileManager">
```
Comment (Ranapop) [inline .idea/inspectionProfiles/profiles_settings.xml]: ignore this file

File: .idea/misc.xml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Ranapop) [inline .idea/misc.xml]: ignore this file

File: .idea/modules.xml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Ranapop) [inline .idea/modules.xml]: ignore this file

File: .idea/vcs.xml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Ranapop) [inline .idea/vcs.xml]: ignore this file

File: data_pipeline/vocab.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 from collections import Counter
```
Comment (Ranapop) [inline data_pipeline/vocab.py:2]: your formatting differs from that of the project -- which makes it seem like there are more changes than there actually are
Please keep the same formatting as in the project and make sure only the relevant changes appear in your PR before opening it

File: models.py [hunk 8]
```diff
@@ -136,12 +138,20 @@ def __init__(self, output_vocab_size: int, config: CfgNode):
...
+    if self.use_pointer_generator:
+        self.p_gen_linear = nn.Linear(config.HIDDEN_SIZE * 3 + config.EMB_DIM, 1, bias=True)
+    self.output_vocab_size = output_vocab_size
   def forward(self,
               input: torch.Tensor,
               previous_state: Tuple[torch.Tensor, torch.Tensor],
               encoder_states: torch.Tensor,
-              attention_mask: torch.Tensor):
+              attention_mask: torch.Tensor,
+              indecies: torch.Tensor,
```
Comment (Ranapop) [inline models.py]: spelled as indices (not indecies)

File: models.py [hunk 9]
```diff
@@ -136,12 +138,20 @@ def __init__(self, output_vocab_size: int, config: CfgNode):
...
+              extended_vocab_size: int,
+              batch_size: int,
+              device):
     """
     Args:
       input: Decoder input (previous prediction or gold label), with shape
```
Comment (Ranapop) [inline models.py:157]: Make sure you also update the comments when adding new parameters

File: train_concept_identification.py [hunk 10]
```diff
@@ -140,26 +151,62 @@ def extract_padding(outputs, eos_index):
...
 
 def indices_to_words(outputs_no_padding,
-                     vocabs: Vocabs):
-  ids_to_concepts_list = list(vocabs.concept_vocab.keys())
-  concepts_as_list = [ids_to_concepts_list[int(id)]\
-                        for sentence in outputs_no_padding for id in sentence\
-                        if ids_to_concepts_list[int(id)] != UNK]
+                     extended_vocab,
+                     config: CfgNode):
+    # TODO put config and use concept_vocab if not pointer generator
```
Comment (Ranapop) [inline train_concept_identification.py]: can't this todo be removed?

**Review by Ranapop — CHANGES_REQUESTED** [review]
I took a more superficial look for now and:
* unnecessary files were pushed
* change in formatting makes reviewing difficult
* some code improvements should be made before a more in depth PR (look over namings, remove unused comments, document functions modified) -- I left some comments in this regard but please take a closer look at the code before requesting a review

File: config.py [hunk 11]
```diff
@@ -14,6 +14,10 @@
 _C.CONCEPT_IDENTIFICATION.LSTM_BASED.HIDDEN_SIZE = 40
 # Number of layers.
 _C.CONCEPT_IDENTIFICATION.LSTM_BASED.NUM_LAYERS = 1
+# Drop-out.
+_C.CONCEPT_IDENTIFICATION.LSTM_BASED.DROPOUT_RATE = 0.6
+# Pointer generator
+_C.CONCEPT_IDENTIFICATION.LSTM_BASED.USE_POINTER_GENERATION = True
```
Comment (alex-frasie) [inline config.py]: Use pointer generaTION? Isn't it Generator?

File: tests/models_test.py [hunk 12]
```diff
@@ -191,11 +196,15 @@ def test_seq2seq_train(self):
     output_seq_len = 5
     input_vocab_size = 10
     output_vocab_size = 20
-    cfg = get_default_config()
+    config: CfgNode
+    config = get_default_config();
```
Comment (alex-frasie) [inline tests/models_test.py]: Feels redundant to declare it and then assign its value to the default config, what was wrong with the previous version? Couldn't you merge the 'opts' to the cfg?

File: data_pipeline/dataset.py [hunk 13]
```diff
@@ -34,6 +36,7 @@ def add_eos(training_entry: TrainingEntry, eos_token: str):
 
 def numericalize(training_entry: TrainingEntry,
                  vocabs: Vocabs,
+                 use_shared: bool,
                  glove_embeddings: GloVeEmbeddings):
   """
   Processes the train entry into lists of integeres that can be easily converted
```
Comment (Ranapop) [inline data_pipeline/dataset.py:61]: Please document the new param (use_shared) in the comment. It would be good to also add a doc for glove_embeddings

File: models.py [hunk 14]
```diff
@@ -151,6 +161,10 @@ def forward(self,
       encoder_states: Encoder outputs with shape
         (input seq len, batch size, hidden size).
       attention_mask: Attention mask (batch size, input seq len).
+      indices: indices in the extended vocab.
```
Comment (Ranapop) [inline models.py]: Indices of what? What are they used for. Indices is a very general word -- and it would be good to document it in more detail for future use

File: models.py [hunk 15]
```diff
@@ -161,17 +175,37 @@ def forward(self,
...
     # before with a NN layer).
+
     classifier_input = torch.cat(
       (previous_embedding, decoder_state[0], context_vector), dim=-1)
     predictions = self.classifier(classifier_input)
-    return decoder_state, predictions
+
+    # generation probability
+    if self.use_pointer_generator:
+      p_gen_input = torch.cat((context_vector, decoder_state[0], lstm_input), 1)
```
Comment (Ranapop) [inline models.py]: who is dimension 1? is it the same with dim=-1? if yes, maybe that would be preferable

File: models.py [hunk 16]
```diff
@@ -223,11 +260,12 @@ def forward(self,
         (num_enc_layers, batch size, hidden size).
       attention_mask: Attention mask (tensor of bools), shape
         (batch size, input seq len).
+      indices: indices in the extended vocab.
```
Comment (Ranapop) [inline models.py]: same comment as above about indices

File: train_concept_identification.py [hunk 17]
```diff
@@ -140,26 +151,67 @@ def extract_padding(outputs, eos_index):
...
+
+     # compute extended vocabulary size
+    extended_vocab_size = len(extended_vocab.items())
+
+    # add in the extended vocabulary the words from the initial input
+    for sentence in unnumericalized_inputs:
+        for token in sentence:
+            if token not in extended_vocab.keys():
+                extended_vocab[token] = extended_vocab_size
+                extended_vocab_size += 1
```
Comment (Ranapop) [inline train_concept_identification.py]: It would be best to move this into a method (maybe in the same script, maybe in utils) that returns the extended vocabulary
You can call it 'construct_extended_vocabulary' or something along those lines. It would be best to also be documented

File: train_concept_identification.py [hunk 18]
```diff
@@ -140,26 +151,67 @@ def extract_padding(outputs, eos_index):
...
+
+    indices = [[extended_vocab[t] for t in sentence] for sentence in unnumericalized_inputs]
+
+    # numericalized_output
+    gold_outputs = torch.transpose(torch.as_tensor([[extended_vocab[word]
+                                                         if word in extended_vocab.keys()
+                                                         else extended_vocab[UNK] for word in sentence]
+                                                        for sentence in unnumericalized_concepts]), 0, 1).to(device)
```
Comment (Ranapop) [inline train_concept_identification.py]: this could also be moved into a documented method -- and perhaps separated into more lines of code (it's quite hard to follow as it is)

**Review by Ranapop — CHANGES_REQUESTED** [review]
The code still needs to be a bit better structured & documented

File: data_pipeline/dataset.py [hunk 19]
```diff
@@ -150,68 +161,93 @@ def __len__(self):
...
     padded_sentences = [
       torch_pad(s, (0, max_sen_len - len(s))) for s in batch_sentences]
+    # Pad initial sentences.
+    padded_initial_sentences = []
+    for sentence in batch_sentences_strings:
+        padded_sentence = copy.deepcopy(sentence)
+        for i in range(max_sen_str_len - len(sentence)):
+            padded_sentence.append(PAD)
+        padded_initial_sentences.append(padded_sentence)
+
```
Comment (Ranapop) [inline data_pipeline/dataset.py]: could be moved into a documented method
~ the same code follows shortly for concepts

File: utils/extended_vocab_utils.py [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+                extended_vocab_size += 1
+    return extended_vocab, extended_vocab_size
+
+
+def numericalize_concepts(extended_vocab, unnumericalized_concepts):
+    """
+    Numericalize the concepts after the new extended vocabulary
+    :param extended_vocab: the vocabulary after which the concepts will be numericalized
+    :param unnumericalized_concepts: the concepts that need to be numericalized
+    :return: the tensor with the numericalized concepts
```
Comment (Ranapop) [inline utils/extended_vocab_utils.py]: Different documentation from other functions

**Review by Ranapop — APPROVED** [review]
I will approve it for now (to not delay other merges further), but code could still be improved (more function extraction and for sure documentation)
